### DISCUSSION TITLE
Badge on NPM packages to show support for ESM entry point

### DISCUSSION THREAD

**Discussion by jonkoops** [npm/feedback#634] [Website] [answered]
Now that ECMAScript modules are widely supported in the NodeJS and JavaScript ecosystem I find myself more often looking explicitly for modern packages that support these modules out of the box. Currently if I look around the NPM site I have trouble to find the right package, as I would first have to dive into the source code of the package to understand if it actually supports ESM.

I'd love to see a badge, just like the one that indicates that a package has type definitions available to indicate support (see mockup below). This way I can instantly see if the package supports ESM.

![A mockup showing the new badge](https://user-images.githubusercontent.com/695720/151658681-75c3fe16-f075-4dd2-9875-1759aa3adf5e.png)

Detecting this would work in a similar manner as the TypeScript badge, by analyzing the contents of the `package.json` file.

**Comment by ljharb** [comment]
Why do you find yourself doing that? node ESM, and transpiled ESM, can import CJS modules just fine; there’s nothing “not modern” about those.

in other words, every CJS package supports ESM.

**Reply by jonkoops** [reply]
> in other words, every CJS package supports ESM.

It depends on what tools you are using, NodeJS can indeed import CommonJS modules in a cross compatible manner. But when I am developing an application using Snowpack it has to do a lengthy pre-processing task to convert CommonJS modules to ESM. CommonJS modules are also hard to analyze at compile time, which makes tree shaking in tools like WebPack impossible for them.

I want to make a conscious choice to avoid using CommonJS modules in my projects for these reasons, so this feature would make it easier to do so.

**Reply by daKmoR** [reply]
very good question I see two main reasons for it

1. when using packages that are intended to be used in the browser there is big difference
    - esm packages you can run directly in the browser
    - cjs packages you always need to transpile using webpack, rollup, esbuild, ...

2. when you need to bundle your code for whatever reason (vs code extension, chrome extension, optimize for the browser, ...)
   - esm is supported by all major bundlers (I think?)
   - cjs means that some tools (for example rollup) may require extra plugins to work with it (and there are caveats in regards to tree shaking)
  
Here are some more personal opinions.  

1. I personally find it easier to read
2. With esm you can be sure that the package did not use/misuse some of the "very dynamic" parts of require

**Reply by targos** [reply]
> I want to make a conscious choice to avoid using CommonJS modules in my projects for these reasons, so this feature would make it easier to do so.

It would only work for packages that have 0 dependencies. An ESM package can depend on CommonJS packages.

**Reply by ljharb** [reply]
If your goal is to highlight things that work better in specific tools - due to, arguably, flaws in those tools’ design - then perhaps npm isn’t the right place to highlight that. Properly working bundlers and build tools can handle CJS and ESM identically well, since they’re identically statically analizable, and CJS’s capabilities are a strict superset of ESM’s.

If your goal is to use things designed for the browser, “ESM support” in no way indicates that, in either direction.

If your goal is to avoid CJS deps for some reason, you’d indeed (as @targos mentions) need to avoid anything that had a transitive CJS dependency too. That’s going to ace out most of the “supports ESM” packages anyways.

**Reply by jonkoops** [reply]
> It would only work for packages that have 0 dependencies. An ESM package can depend on CommonJS packages.

That is a really good point actually, I had not yet considered that. Being able to tell for even just for the package itself would already be a huge boon I think.

**Reply by daKmoR** [reply]
> If your goal is to avoid CJS deps for some reason, you’d indeed (as @targos mentions) need to avoid anything that had a transitive CJS dependency too. That’s going to ace out most of the “supports ESM” packages anyways.

yes, finding those packages that are valuable to me is the goal of the npm catalog right?

**Reply by jonkoops** [reply]
> If your goal is to highlight things that work better in specific tools - due to, arguably, flaws in those tools’ design - then perhaps npm isn’t the right place to highlight that.

I mean to call this a 'design flaw' is going a bit far IMHO, one could argue the other way that the design of CommonJS is flawed as well. But this all depends on what you need to do with the specific package you are interested in.

I just want to be able to make an informed decision when shopping around for a new package to use to make sure it fits my uses cases.

**Reply by ljharb** [reply]
@daKmoR no, because there's far too many subjective measures of value for it to be appropriate for npm to "bless" any but the most universal/common/uncontroversial.

@jonkoops sure, if we were discussing module systems; but the module systems exist, thus, tooling to deal with them must be judged on how well it does so.

**Reply by jonkoops** [reply]
> tooling to deal with them must be judged on how well it does so.

I think a lot of the frustrations people experience with JavaScript tooling is due to how complex it has become over time. We are a long way away from the 'old days' where you could simply load a script and get on with your work. I personally think that ESM is a great step forward to address this issue.

So if I can use a tool that can provide me a better developer experience and reduce the burden on maintainers of that tooling or packages in general if I only use ESM, then why would I not?

**Reply by ljharb** [reply]
I have tons of reasons why not, but this isn't the place to debate that; the main thing is that it wouldn't be appropriate imo for npm to "bless" something like this when it's still a vanishingly small percentage of the ecosystem that's adopted that approach.

**Reply by justinfagnani** [reply]
> you’d indeed (as @targos mentions) need to avoid anything that had a transitive CJS dependency too

Yes, the best version of this feature would show the badge for packages that are JS modules and have only JS module dependencies.

**Reply by jonkoops** [reply]
@ljharb I don't think this would be a 'blessing' from NPM that says it's somehow superior (again this is all relative to what your use case is). It's about transparency towards the consumer of the package so they can make an educated decision.

**Reply by ljharb** [reply]
(CJS modules are JS modules)

@jonkoops how could npm actually know this tho? Module and Script are not unambiguous parse goals. Are you suggesting npm would read the "exports" field (which has MANY different permutations of feature support depending on node version) and figure out if it declares ESM files, and also not CJS files? What about the non-standard "module" field, or the "jsnext:main" field?

This just isn't something that npm can possibly know with certainty.

**Reply by justinfagnani** [reply]
^ Exactly. The TS tag isn't saying that TS is better, just that the types are available.

**Reply by ljharb** [reply]
@justinfagnani the ask wasn't "tag packages that ship native ESM with a badge", it was "tag packages that ship ESM and do NOT ship CJS". That is a value judgement.

**Reply by justinfagnani** [reply]
This feature needn't support every legacy and non-standard way of declaring standard JS module entrypoints. It only needs to support the `"type"` and `"exports"` fields. Packages that care about displaying this badge can update.

**Reply by justinfagnani** [reply]
@ljharb 
> "tag packages that ship ESM and do NOT ship CJS"

Sorry, I don't see that request here. Just to tag modules that export standard JS modules.

**Reply by ljharb** [reply]
@justinfagnani you're right, it's not in the OP; but it is in this thread.

again, "standard JS modules" is a misnomer. If you're looking to tag packages that claim to export ES Modules (the term it's known by), then npm would have to parse the "exports" field, and would have to understand multiple permutations of support for it (because there exist many different featuresets across all the node versions, and npm can't just support "node latest").

**Reply by ljharb** [reply]
I will add, though, that knowing a package contains ESM exports (on just the main? on other entry points? what if the main is CJS but there's an ESM entry point?) tells you nothing about whether it will work in the browser, with or without bundling.


The "types" badge tells you whether it will work with TS, full stop. This "ESM" badge wouldn't tell you anything except what's supposed to be a hidden implementation detail of the package: the module format.

**Reply by justinfagnani** [reply]
I prefer to call them standard JS modules because that's what they are. We don't call other standard JS features ES-something. I use functions, not ES functions, Promises, not ES Promises. Anyway, you obviously get my meaning, so...

I still don't think that this feature would have to support permutations of the `"exports"` field. If Node has finally settled on the format for `"exports"` (and they're deprecating and removing the old format), then this feature absolutely could support only the final format. Again, packages that care about the badge could update.

**Reply by ljharb** [reply]
Would this only apply to the main export? What if i had a package that offered an ESM entry point, but not under the main/dot?

Would this mean that to get the badge, I'd have to add an `index.mjs` file to all my packages that did `import x from './index.js'; export default x`? That seems like it'd be a perverse incentive (adding filesize weight to my 300+ packages) that wouldn't end up adding any value.

**Reply by justinfagnani** [reply]
Don't those questions apply to the TypeScript badge too? The typings could be wrong or only for a subset of entrypoints.

And if you wanted to lie to your potential users and piss them off... I guess that's one strategy.

**Reply by ljharb** [reply]
It wouldn’t be a lie.

the types badge shows up for DT packages too, so I’m not forced to add unnecessary bloat to my packages to get the badge - so there’s no perverse incentive for the types to be wrong.

**Reply by justinfagnani** [reply]
I don't know why you would go out of your way to get the badge if you don't think it has value.

**Reply by ljharb** [reply]
Because people aren’t rational consumers, and their choices are motivated by shiny things.

Why do you think it would be a lie tho? Is it perhaps because you actually want the badge to mean “no CJS”?

**Reply by justinfagnani** [reply]
I want it to mean available transitively as standard JS modules. That doesn't mean that there can't also be parallel CJS entrypoints as `"exports"` allows.

**Reply by ljharb** [reply]
Then my suggestion satisfies that meaning.

**Reply by jonkoops** [reply]
> how could npm actually know this tho?

I'd say that we have a pretty solid way of telling if a package is written in ESM, the NodeJS [documentation](https://nodejs.org/api/packages.html) is pretty explicit about what qualifies.

> the ask wasn't "tag packages that ship native ESM with a badge", it was "tag packages that ship ESM and do NOT ship CJS". That is a value judgement.

Although I don't hide that it's my personal opinion that we should strive to eliminate CommonJS for the sake of reducing the complexity of the ecosystem in the long run, that is not what I am stating here. I never advocated that packages should be 'ESM only' for them to receive this badge, only that they ship with support for ESM as outlined in the NodeJS documentation.

> The "types" badge tells you whether it will work with TS, full stop. This "ESM" badge wouldn't tell you anything except what's supposed to be a hidden implementation detail of the package: the module format.

I am sorry but although a lot of effort has been put into interoperability between CommonJS and ESM in runtimes (which is appreciated), they are simply not compatible with each other everywhere. It's great that sometimes this is abstracted for you in a way to lessen the pain, but it is still there in tooling outside of NodeJS.

> Because people aren’t rational consumers, and their choices are motivated by shiny things.

This is a bit derogative. I think there are valid reasons to want to pick a package in the ESM syntax as mentioned before.

**Reply by ljharb** [reply]
The reasons I’ve heard - which i obviously don’t agree with, but that’s immaterial - are about trying to avoid CJS in your dependency graph. Because of dependencies, a badge wouldn’t achieve that goal. Because ESM can always import CJS (it’s the reverse that’s problematic), almost every package on npm is compatible with ESM.

Can you give me an example of a package that you would not want to have this badge, that is compatible with require/CJS but not with ESM, and for which transitive dependencies meeting the same criteria wouldn’t defeat the goal?

**Reply by jonkoops** [reply]
> Can you give me an example of a package that you would not want to have this badge, that is compatible with require/CJS but not with ESM, and for which transitive dependencies meeting the same criteria wouldn’t defeat the goal?

I am not going to link individual packages here as the NodeJS documentation already provides what I would consider an accurate description of a package that does not support an ESM entry point.

As for the transitive dependencies, I can already browse a package's dependencies trough the site. So if I can see that these dependencies ship ESM there I am already more happy than I was before (as I don't have to browse the source of each package).

Again, I am not advocating 'pure ESM' packages here with 'pure ESM dependencies'. But rather the ability to see if a specific package exposes in a standardized manner (not simply includes syntax) an ESM compatible entry point.

A badge with the title text like "This package exposes an entry point that supports ECMAScript modules" would be enough.

**Reply by ljharb** [reply]
@jonkoops right, but "a CJS file" is an entry point that supports ECMAScript modules, because you can import them (which is the only definition of "support" that i think anyone should care about).

**Reply by justinfagnani** [reply]
For examples, `preact` should have the badge, `react` shouldn't.

**Reply by jonkoops** [reply]
I've updated the title to be a bit more clear that this is not about 'ESM Only' packages.

**Reply by ljharb** [reply]
@justinfagnani thanks. can you elaborate on why? Both can be `import`ed in native ESM modules. Both require bundling/building (like every CJS and ESM package) to work in all browsers (especially since import maps is not yet a web standard).

**Reply by justinfagnani** [reply]
React absolutely cannot be imported into standard JS modules. It needs to be transformed into another format first.

Preact does not require bundling, you can use it right from CDNs like unpkg.

**Reply by ljharb** [reply]
In node - the primary thing npm should be caring about - it absolutely can be.

Preact is a rare case where it either has no dependencies, or, it provides a bundle already that includes all of them. Are you suggesting you want an ESM badge on packages that meet _this_ criteria?

**Reply by justinfagnani** [reply]
npm is used by browser packages too.

**Reply by ljharb** [reply]
Absolutely it is! But it’s not the primary consumer, and the implementation-defined parts of ES Modules that browsers have chosen to implement (URLs, for example, are non-standard - the standard only has “string specifiers”) aren’t the only implementation of ESM.

If something like this is to be added, it’s goals and criteria have to be very precise, or else it’s technically unfeasible.

If the goal is to answer “can this package be imported”, then every CJS package already qualifies. If the goal is “can this package be imported in ESM in browsers without a build process”, though, then that’s a question that isn’t *possible* to statically answer, and arguably that’s not a question npm *should* be trying to answer (*which* browsers/versions, for example, is an immediate question that would need to be answered).

**Reply by jonkoops** [reply]
> right, but "a CJS file" is an entry point that supports ECMAScript modules, because you can import them (which is the only definition of "support" that i think anyone should care about).

Again, in the NodeJS ecosystem this might be true, but outside this is not always the case. For example:

- Webpack cannot [tree-shake](https://webpack.js.org/guides/tree-shaking/) CommonJS modules
- Rollup cannot bundle CommonJS without a [plugin](https://www.npmjs.com/package/@rollup/plugin-commonjs)
- Browsers don't understand CommonJS modules
- Deno [does not support](https://deno.land/manual@main/npm_nodejs) CommonJS modules

**Reply by jonkoops** [reply]
I don't think the proposed implementation in any way suggests that browsers could directly run a package that uses ESM (although a lot less transformation and logic would be necessary than with CommonJS).

This is simply about whether a package (and not even it's dependencies) exposes an entry point to an ESM module. Which can be relevant for those who are not using a NodeJS environment.

**Reply by ljharb** [reply]
Right - but that will create a perverse incentive for packages to unnecessarily add an ESM entry point that re-exports their existing CJS implementation.

I could understand two badges - “can be required” or “can be imported” - and most packages will have both, but ESM-only packages will have just one, and C++ packages will have just one.

**Reply by jonkoops** [reply]
That's a valid concern. I am sure there are those out there that might be using tooling that can only deal with CommonJS, and don't want to accidentally install an ESM version.

I agree that adding both a badge for CommonJS and ESM to give people a transparent picture is the better option 👍

**Reply by ljharb** [reply]
To be clear, I’m suggesting a badge for “requireable” and “importable”. The module format shouldn’t matter, and won’t actually give you the property you want, and is very difficult for npm to determine.

**Comment by rauschma** [comment]
I like how Skypack handles this – e.g. (details on the right): https://www.skypack.dev/view/preact

**Reply by jonkoops** [reply]
I love how Skypack shows their package score, it is really helps understand the quality in a clear overview. I've always found it very hard to understand the seemingly arbitrary 'Quality' score on the NPM site.

**Reply by rauschma** [reply]
I don’t like the score itself (impossible to condense this into a single number), but the feature list is nice.

**Reply by jonkoops** [reply]
Yes, I like the transparency that the list provides. The quality here is relative to what you need as a consumer, but in the case of Skypack it reflects what makes it easier to provide their service.

**Comment by Westbrook** [comment]
@jonkoops really interesting request, could definitely see it being useful in a lot of contexts. Wonder if you'd share some additional information about your usage on "supports ES Modules" as "modern". Using JS with CJS modules in a node environment, wouldn't a project be able to get the "modern" label just as easily as if it were shipped as ES Modules based on that actual JS that it contained? I definitely have personal feeling in support of ES Modules as a whole, and am interested in your thoughts on whether a JS library could be "modern" without them.

What's more, I ask, because I see "has ES Modules" as only half the story of a bigger request in my work, and that's "is browser ready". This takes into account things like CJS modules or `process.env` and others not working in the browser by default and thus requiring build tools to leverage in that context. Some other commenters point to the very correct point that any number of tooling situations could bridge that gap between Node first dependency and browser capable dependencies. However, a clear indicator that you could forego the hard part of setting up advanced tooling while getting into a new project as indicated by NPM could do wonders for how new developers got started in the craft.

With that in mind, I'd much rather see NPM take the next step in its evolution from Node (only) Package Manger to de facto JS, or even _everything_, Package Manager and more clearly define the contexts where in a listed package can be used. "Is browser ready" might not even be the only context for that, especially as more and more tools make it into NPM. It might make sense to start clarifying more and more of these contexts (wasm, node, CSS, there was talk of C++ above) so that NPM could support getting users to success even faster than ever.

I look forward to more on your thoughts here, and strongly agree that NPM is in a great place to leverage its outsized and growing influence in this area to support a demystification of the ecosystem on any number of axes... +1!

**Reply by Seanmclem** [reply]
For me it's Deno. I love how quickly and easily I can use any esm package and like to avoid the added difficulty of not using them

**Reply by jonkoops** [reply]
> Wonder if you'd share some additional information about your usage on "supports ES Modules" as "modern".

> I definitely have personal feeling in support of ES Modules as a whole, and am interested in your thoughts on whether a JS library could be "modern" without them.

So I think it's important here to note that I only consider it to be "modern" because it is a newer standard. However whether this is suitable for your needs is entirely up to the tooling you use.

> Using JS with CJS modules in a node environment, wouldn't a project be able to get the "modern" label just as easily as if it were shipped as ES Modules based on that actual JS that it contained?

I don't quite understand what you are asking here, could you elaborate?

> What's more, I ask, because I see "has ES Modules" as only half the story of a bigger request in my work, and that's "is browser ready". 
 
This is true, just because a module is written in ESM does not mean that it cannot be used inside of the browser. Some assumptions that exist in Node (like importing a package by it's name) simply do not (yet) exist in the browser. Having ESM in a package does however allow tooling to further optimize it when building, which is essentially why I would want this 'ESM' badge.

Perhaps the [`browser`](https://docs.npmjs.com/cli/v8/configuring-npm/package-json#browser) field can be used with [conditional exports](https://nodejs.org/api/packages.html#packages_community_conditions_definitions) to accomplish this.

**Reply by ljharb** [reply]
No, it doesn’t; there’s nothing about ESM that’s more optimizeable than CJS - it’s just that a number of tools don’t bother to implement the same optimizations for CJS modules.

**Reply by jonkoops** [reply]
> No, it doesn’t; there’s nothing about ESM that’s more optimizeable than CJS

CommonJS is not completely statically analyzable due to the fact it it operates at runtime, this is a fact. Some tools depend on this to optimize code beyond what CommonJS can accommodate.

**Reply by ljharb** [reply]
@jonkoops both CJS and ESM have static and dynamic forms. It is a fact that they are identical in this way - dynamic import for ESM, and require with “not a string literal” for CJS, are both impossible to analyze; static import for ESM and require with a string literal are both identically static.

**Comment by mhdawson** [comment]
> I’m suggesting a badge for “requireable” and “importable”

This makes sense to me.

**Reply by kito99** [reply]
We really need something like this. If someone is looking for an ES module, right now it's way too much work to determine if it's supported, especially for someone new to the ecosystem.

**Comment by MylesBorins** [comment, marked as answer]
There's a historic discussion about identifying a number of format details of a package, does it make sense to folks this conversation into there?

https://github.com/npm/feedback/discussions/13

**Reply by Westbrook** [reply]
That’s quite interesting. Have you seen much motion at the project level from that conversation?

**Reply by jonkoops** [reply]
I think this overlaps the previous discussion quite well, if there is some way the specific concerns of this request can be folded into that larger discussion I see no problem with that.

**Reply by bennycode** [reply]
Is it difficult to build a proof-of-concept that shows an ESM badge when `"type": "module"` is found in the `package.json`? That may provide important feedback to drive #13.

**Reply by ljharb** [reply]
@bennycode type module doesn't necessarily, or exclusively, mean "ESM", though.

**Reply by jonkoops** [reply]
FYI the [Yarn site](https://yarnpkg.com/package?q=vite&name=vite) now does something similar, it considers the presence of an entry-point the qualifier.

![Screenshot of the Yarn site showing this feature](https://github.com/npm/feedback/assets/695720/b3e57482-8073-4b9c-8bec-ef6a8fca130d)

**Reply by ljharb** [reply]
Sadly that only checks format and not module system compatibility :-/

**Reply by jonkoops** [reply]
Yeah, but I think that is good enough. I think the phrasing here is good, if the package author specifies a specific entry-point that is reflected as supported here. This is a case of don't let perfect be the enemy of good.

**Reply by justinfagnani** [reply]
The problem is that it isn't sufficient. If the package has non-standard-module dependencies, then it will still fail to load in browsers.

For instance, I recently saw that the `random` package is now published as modules, but it depends on `seedrandom` which isn't, so it wouldn't load correctly from module CDNs.

I think that case is the more important one for a badge. Those using Node (after `require(esm)` is available at least) or bundlers in dev don't have to worry about the format so much. Those loading modules directly do, but they need modules to be standard _deeply_.

**Reply by jonkoops** [reply]
> Those loading modules directly do, but they need modules to be standard deeply.

That could be a qualifier for showing the badge. Or it could be semi-transparent with a tooltip message like "This package provides an ESM entry-point, but some of it's dependencies do not."

**Reply by jonkoops** [reply]
@MylesBorins is there a way to actually progress this issue? I'd love to contribute something to the NPM site, but considering the fact that AFAIK it's closed source, there seems to be very little the community can do to enhance the experience there. Not just talking about this specific issue, but also various other enhancements.

**Reply by ljharb** [reply]
Indeed, as @justinfagnani mentions, the actual question that matters to users is "can i use this package in my app?"

The module format may or may not matter; whether it works in browsers and/or servers may or may not matter; how large it is may or may not matter; how many deps it has may or may not matter.

"Perfect is the enemy of the good" doesn't handwave away that only talking about module formats *isn't good* - it's just not useful information by itself, and it's more likely to be harmful (ie, perverse incentives) than helpful.

**Reply by jonkoops** [reply]
> "Perfect is the enemy of the good" doesn't handwave away that only talking about module formats isn't good - it's just not useful information by itself, and it's more likely to be harmful (ie, perverse incentives) than helpful.

@ljharb we fundamentally disagree on this, and I don't think we will or have to settle our differences here. It might not matter to you, but that doesn't mean it doesn't matter to others. I think that is pretty evident from this being the [second most upvoted discussion](https://github.com/npm/feedback/discussions?discussions_q=sort%3Atop+), and from other reaching out here to discuss this feature request.

Not talking about a problem because you don't want it to exist in the larger ecosystem outside of Node.js isn't going to make the problem go away. All I am asking for here is the information that might be relevant in making such a choice is presented to the user. so they can make a judgement themselves.

**Reply by jogibear9988** [reply]
I've found so much packages, wich for example have a ESM entry point, but their internal imports do not have file extensions, so the package does not work without using a bundler.
I try this for example most of the time in my web-component-designer, wich could load directly npm packages. Nearly for every package I try I need to create bug reports or fixes, cause the package itself or some of it's dependecies has errors in the ESM exports.

**Reply by justinfagnani** [reply]
> I've found so much packages, wich for example have a ESM entry point, but their internal imports do not have file extensions, so the package does not work without using a bundler.

At least those will work with import maps or a simpler transform that only resolves import specifiers.

But it does raise an important and subtle point: It may be obvious to many that using npm packages names requires an import map or transform for any module that imports an npm dependency. It may be less obvious that using non-relative paths (including paths without extensions) internally requires an import map or transform as well. And it may be even less obvious that even though both cases require an import map or transform, needing import map entries for _internal_ specifiers can drastically blow up the size of the import maps compared to just needing 1 or 2 entries per package.

So when determining whether a package is "web ready", it matters what requirements are made of the specifiers, ranging from only being valid URLs or relative paths (not practical, IMO); to allowing bare specifiers for package names, but otherwise having paths that don't require further resolution (practical _and_ optimal for import maps, IMO, but commonly violated), to allowing any specifier that can be resolved by Node resolve (pretty much rules out import maps).

I suspect there would be a fair amount of disagreement about which specifier requirement would make it into a "web ready" designation.

**Comment by jogibear9988** [comment]
> > I've found so much packages, wich for example have a ESM entry point, but their internal imports do not have file extensions, so the package does not work without using a bundler.
> 
> At least those will work with import maps or a simpler transform that only resolves import specifiers.

Yeah for sure, but I could only create them automatically when the package.json has a "exports" section (wich most of them do not have atm.) where these all are listed. If not, it may be possible for a bundler or other local app wich parses the files and searches for the imports, but not for my lib wich only loads the npm's (the other also would be to slow to do live)

**Reply by jonkoops** [reply]
@jogibear9988 probably best to reply in-thread, rather than starting a new one, otherwise this discussion will become very disconnected.

**Reply by jogibear9988** [reply]
yeah, was an accident
